sql/core/src/main/scala/org/apache/spark/sql/Melt.scala

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.spark.sql.catalyst.analysis.{AnsiTypeCoercion, TypeCoercion}
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Cast, Expression, Literal}
+import org.apache.spark.sql.catalyst.expressions.objects.AssertNotNull
+import org.apache.spark.sql.catalyst.plans.logical.Expand
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{DataType, StringType}
+
+private[sql] object Melt {
```

Comment:
I think this whole class has to be a separate `SparkPlan` ..

Reply:
That would be ideal, because `Melt` than becomes explicit in the Spark plan. Since the implementation is completely based on a single `Expand`, how could that `Melt` reuse the `Expand` implementation? A `case class Melt(...)` cannot extend `case class Expand`.

In `SparkStrategies.scala`, the logical `Melt` could be replaced by the respective execution `ExpandExec`, but then there would be no explicit `MeltExec`.

Reply:
I could make that `Melt` add the respective `Expand` as its children node and either have a no-op `MeltExec` or have the `Melt` removed from the execution plan. What do you think?

Reply:
Alright, `Melt` is now a `LogicalPlan`, which already fixed / implemented SPARK-39292.